crates/subspace-networking/src/node_runner.rs

```diff
@@ -364,7 +373,16 @@ where
                 };
                 debug!("Connection closed with peer {peer_id} [{num_established} from peer]");
 
-                shared.connected_peers_count.fetch_sub(1, Ordering::SeqCst);
+                shared
+                    .kademlia_tasks_semaphore
+                    .shrink(KADEMLIA_CONCURRENT_TASKS_BOOST_PER_PEER)
+                    .await
+                    .expect("Failed to shrink kademlia_tasks_semaphore");
```

Comment:
This message is not a proof, see https://subspacenetwork.slack.com/archives/C03R16SFFJM/p1668314727705509 I shared a while ago about how to approach it.

Reply:
I moved the panic() inside the Sem implementation. In this case, this is not an external API where we should fail the request instead of crashing. The panics happen in these two cases: more permits are freed than allocated, capacity is shrunk more than current value.

Both these would be bugs, I think we should crash the binary and debug, instead of continuing in the bad state which could lead to other undefined behavior. WDYT?